### DISCUSSION TITLE
use tree-shaking to reduce the bundle size by >90%

### DISCUSSION THREAD

**Discussion by samuelstroschein** [sinclairzx81/typebox#614] [Ideas]
## Problem 

Typebox ships too large bundle sizes for resource constraint environments. Aka typebox can't be used.

The following code ships 36.3kb [[source](https://bundlejs.com/?q=%40sinclair%2Ftypebox&treeshake=%5B*%5D&text=%22export+type+Settings+%3D+Static%3Ctypeof+Settings%3E%5Cnexport+const+Settings+%3D+Type.Object%28%7B%5Cn%5CtstoragePath%3A+Type.String%28%29%2C%5Cn%7D%29%22&config=%7B%22analysis%22%3Atrue%7D)]

```ts
 import { Type, type Static } from "@sinclair/typebox"

export type Settings = Static<typeof Settings>
export const Settings = Type.Object({
	storagePath: Type.String(),
})
```

## Proposal

Leverage ESM tree-shaking to decrease bundle sizes by over 90%. 

For comparison, [valibot](https://valibot.dev/), which leverages tree-shaking, bundles 703 bytes for the same code [[source](https://bundlejs.com/?q=valibot&treeshake=%5B*%5D&text=%22export+type+Settings+%3D+Output%3Ctypeof+Settings%3E%5Cnexport+const+Settings+%3D+object%28%7B%5Cn%5CtstoragePath%3A+string%28%29%2C%5Cn%7D%29%22&config=%7B%22analysis%22%3Atrue%7D)]. A bundle size reduction of 98% compared to typebox. 

### Examples

The following code would only import `Object` and `String` validators without all available `Type.*` stuff. 

```ts
import { Object, String, type Static } from "@sinclair/typebox"

export type Settings = Static<typeof Settings>
export const Settings = Object({
	storagePath: String(),
})
```

The syntax of `Type.*` could even be sustained with namespace imports which are tree-shaken by bundlers too. 

```ts
import * as Type from "@sinclair/typebox"

export type Settings = Static<typeof Settings>
export const Settings = Type.Object({
	storagePath: Type.String(),
})
```

## Additional information 

The `Value` import of typebox is too large as well, counting >70kb that are included in the bundle size. Valibot's equivalent `parse` bundles only 100 bytes.

**Comment by sinclairzx81** [comment]
@samuelstroschein Hiya,

---

### Bundle Sizes

Just for reference, the following table are the current bundling sizes for various libraries.

| Type System | Minified | GZipped   | Reference |
| ---         | ---      | ---       | ---                                                                |
| TypeBox     | 34.3kB   | 8.3kB     | [0.31.17](https://bundlephobia.com/package/@sinclair/typebox@0.31.17) |
| Valibot     | 35.5kB   | 6.6kB     | [0.18.0](https://bundlephobia.com/package/valibot@0.18.0) |
| Zod         | 57.0kB   | 13.2kB    | [3.22.2](https://bundlephobia.com/package/zod@3.22.2) |
| Ajv         | 119.6kB  | 35.2kB    | [8.12.0](https://bundlephobia.com/package/ajv@8.12.0) |
| HyperJump   | 44.3kB   | 11.6kB    | [1.6.1](https://bundlephobia.com/package/@hyperjump/json-schema@1.6.1) |
| ArkType     | 67.7kB   | 18.9kB    | [1.0.21-alpha](https://bundlephobia.com/package/arktype@1.0.21-alpha) |

Note that the TypeBox result above relates specifically to the TypeBuilder (which is currently non-shakable). 

---

### ESM Modules

The TypeBox TypeBuilder (everything inside `typebox.ts`) is long overdue for a module split. This module has been prepared for splitting over the past several iterations of the library, but currently pending "the split" until things are closer to 1.0. 

The anticipated import scheme post split will be:

```typescript
import Type, { Static } from '@sinclair/typebox'
```
Given this is a breaking change, there is some reluctance to enact this change until everything is finalized for 1.0 (and that this scheme will be supported into the future). So this is planned for, just not for a while.

### Value Modules

> The Value import of typebox is too large as well, counting >70kb that are included in the bundle size. Valibot's equivalent parse bundles only 100 bytes.

You won't get to 100 bytes, but you can significantly reduce the size of the Value modules by importing each function individually. 

```typescript
import { Check } from '@sinclair/typebox/value/check' // > 24 kb
```
Note: TypeBox does need to import logic to check for all types (as it can't know in advance what types are going to passed to it). Work was done on the 0.30.x and 0.31.x revisions to reduce output size (many code level optimizations), as well as to make each function with it's associated infrastructure individually importable.

Outside of the above, there is very little else that can be done while keeping schematics and validation logic separated (which is a key design principle for the library). However, there may be room to explore more optimal (bundler friendly) Json Schema validation libraries external to TypeBox. 

It would be interesting to see just how minimal a Json Schema implementation could get.

Hope this brings some insight
Cheers
S

**Reply by samuelstroschein** [reply]
As always, thanks for the in-depth reply. 

> The anticipated import scheme post split will be:
> import Type, { Static } from '@sinclair/typebox'

How is `Type` tree-shakable with this import scheme?

> Just for reference, the following table are the current bundling sizes for various libraries.

The bundle sizes of libraries are irrelevant if tree-shakable. As it stands right now, importing `Type` leads to 36.3kb that are bundled. I fear that the bundling size with the import schema your propose will make no significant differences as all `Type` annotations are still bundled.

**Reply by sinclairzx81** [reply]
@samuelstroschein Hiya,

I would expect something like the following (with the default export and type functions would be statically analyzable by bundle optimizers)

```typescript
import Type from 'typebox'
// or ...
import { String, Object } from 'typebox'
```
> The bundle sizes of libraries are irrelevant if tree-shakable. As it stands right now, importing Type leads to 36.3kb that are bundled. I fear that the bundling size with the import schema your propose will make no significant differences as all Type annotations are still bundled.

Keep in mind, individual imports (i.e. as per Valibot) are actually not a big priority for TypeBox, rather the main motivation is to adopt ESM (in whatever capacity) and ensure bundlers have enough to statically analyze (and optimize) the bundle.

In this regard, I feel the `default` export is generally how people would prefer to consume TypeBox under ESM (vs the barrel import) as the default narrows the import scope to only `Type.*`. Additional infrastructure (including the Types themselves) can be imported via `{ ... }` which I would expect a capable bundle optimizer to understand no default import was used (so can be omitted from the output bundle)

This is the reasoning anyway. It could use some testing under popular optimizers.

**Reply by samuelstroschein** [reply]
I am implementing a tree-shakable i18n library. My tests with rollup/vite yielded the following result:

```ts
// not tree-shakable 
import Type from 'typebox'

// tree-shakable
import { String, Object } from 'typebox'

// syntactic sugar of the tree-shakable version is a namespace import
// bundlers ignore the syntactic sugar and treat the following import as 
// import { Object } from "typebox"
import * as Type from "typebox"
Type.Object()
```

Under any circumstance, importing `{ Type }` or `Type` as default import will lead to (too) large bundle sizes. To avoid a breaking change, you can keep `import { Type }` but deprecate `Type`, expose `import { Object, String }` and mention in the docs that a namespace import can be used `import * as Type` for a nicer DX.

**Reply by samuelstroschein** [reply]
I am pressing the issue because typebox is too large for [inlang plugins](https://inlang.com/marketplace) where a plugin has a size of <10kb. Using typebox in its current form would add 36kb to each plugin. A no go. Hence, we had to switch to valibot for plugins.

**Reply by sinclairzx81** [reply]
@samuelstroschein Heya,

The following would be ok though right?

```typescript
// not tree-shakable 
import Type from 'typebox'

// tree-shakable
import { String, Object } from 'typebox'

// presumed also tree-shakable
import * as Type from 'typebox'
```
> I am pressing the issue because typebox is too large for [inlang plugins](https://inlang.com/marketplace) where a plugin has a size of <10kb. Using typebox in its current form would add 36kb to each plugin. A no go. Hence, we had to switch to valibot for plugins.

Inlang looks like a cool project btw. And yeah, that's a bit unfortunate about the plugin size constraints. The TypeBuilder currently gzip's down to around `8kb` with everything in there (assuming you don't need validators). With validators, the minimum you would be looking at around 24kb. Even with TB opting for barrel imports, you're still going to be looking at 20-30kb for validators and additional TB components.

I'm not very keen to deprecate `Type.*` (as I feel this actually improves DX as it narrows specifically to constructible types vs the [other infrastructure](https://github.com/sinclairzx81/typebox/blob/master/src/typebox.ts#L1018-L2828) used for type composition). Explicit imports of types via `{ String, Object }` will most likely be supported via ESM and permit developers to optimize if they need that. I feel this strikes a good balance between something like Zod (not tree shakeable) and Valibot (by being optionally tree shakeable) while still remaining distinctly "TypeBox" (as well as not relying too much on the capabilities of JS optimizers)

But let me give it some thought. ESM work is still a fair way off (I have some prototypes I need to test against). Also, I wouldn't expect ESM before the end of the year (but will see how the workload is in December). Just keep in mind that implementing any of this will be significant breaking API change, so do want to have good justification and rationales in place before going ahead with anything. It's a bit of a balancing act (and there are other trade-offs to consider), but should deprecation of `Type.*` be desirable (when contrasted with [other functionality](https://github.com/sinclairzx81/typebox/blob/master/examples/typedef/typedef.ts)), I'll give it a review when I get around to the work.

Let's keep this thread active and open for discussion :)

**Reply by samuelstroschein** [reply]
@sinclairzx81 

> The following would be ok though right?

```ts
// not tree-shakable 
import Type from 'typebox'

// tree-shakable
import { String, Object } from 'typebox'

// presumed also tree-shakable
import * as Type from 'typebox'
```

I would remove the default import. It seems to me that you can implement tree-shaking in a non-breaking change manner as follows: 

```ts
// current import schema, still supported
import { Type, Static } from '@sinclair/typebox'

Type.String() // -> warning that `Type` is deprecated in favor of `import * as Type` but it still works!

// new exports of individual types (tree-shakable) 
import { String } from  '@sinclair/typebox'

String()

// syntactic sugar for type 
import * as Type from "@sinclair/typebox"

// really a minor "issue"/non-issue that static now runs under type. 
type X = Type.Static<typeof X>
const X = Type.String()

```

The code above is non-breaking while encouraging best practices and automatic tree-shaking, making typebox substantially better (and hopefully enabling new use cases for typebox in resource constraint environments). No developer would have to debug "why is typebox shipping 30k+ kb" if you deprecate the direct import of Type and do not expose a default import.

**Comment by samuelstroschein** [comment]
Zod released a tree-shakable library https://v4.zod.dev/packages/mini. 

They used the pattern I described in https://github.com/sinclairzx81/typebox/discussions/614#discussioncomment-7184707 with `import * as X`. The hesitation to use TypeBox because of the lack of tree-shaking is high. Our use case revolves around plugins that bloat from <10kb to >40kb if TypeBox is used. 

A non-breaking implementation could be a sub-path. 

```
import * as T from "@sinclair/typebox/tree-shakable"
```

**Reply by sinclairzx81** [reply]
@samuelstroschein Hello, 

This discussion is quite old (late 2023) and TypeBox has since gone ahead with optimizing the codebase for splitting (early 2024 iirc). According to bundlejs, the following code minifies and gzips to 1.04 kB

```typescript
export const T = Type.Object({
  x: Type.Number(),
  y: Type.Number(),
  z: Type.Number(),
})

// require import here as share link breaks code
import * as Type from "@sinclair/typebox@0.34.33"
```

Reference Link (Click on Build)

https://bundlejs.com/?text=%22export+const+T+%3D+Type.Object%28%7B%5Cn++x%3A+Type.Number%28%29%2C%5Cn++y%3A+Type.Number%28%29%2C%5Cn++z%3A+Type.Number%28%29%2C%5Cn%7D%29%5Cn%5Cn%2F%2F+require+import+here+as+share+link+breaks+code%5Cnimport+*+as+Type+from+%5C%22%40sinclair%2Ftypebox%400.34.33%5C%22%22

Does this solve the issue?

**Reply by samuelstroschein** [reply]
Oh yes it does! 

I haven't seen the `import * as Type` mentioned in the docs, nor any reference to tree-shaking. Updating the docs would help!
